### PR TITLE
More SAO related changes (revised)

### PR DESCRIPTION
**Pull request by sfan5**
Various bug & documentation fixes related to entities.

**DO NOT SQUASH**

## To do

This PR is a Ready for Review.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
Can you please elaborate what this PR is good for?
This this just improved logging or is there more?

**Comment by Wuzzy2** [comment]
>     * The object is removed after returning from Lua, however the `ObjectRef`
>      instantly becomes invalid and unusable.

What does this mean, “unusable” and “invalid”? What happens when a rogue mod tries to use the object anyway? Does the ObjectRef magically become `nil`? Will Minetest go “boom” when I try to use it anyway? Will a warning be written? Etc. And what can the modder do to defend against possible bugs?

Whatever the answer is, this text needs improvement as I don't fully understand it.

**Comment by sfan5** [comment]
> This just improved logging or is there more?

Yes, the rest is random bugfixes and other things that were necessary.

> What does this mean, “unusable” and “invalid”?

It means all methods return `nil` without doing anything.

> Will a warning be written?

I briefly looked into that but that wasn't easy to do in the code. Calling object methods after `remove()` is an easy to spot bug anywayn.

> And what can the modder do to defend against possible bugs?

Possible bugs where?

**Comment by Wuzzy2** [comment]
> It means all methods return nil without doing anything.

Ah. This definitely needs to go in the documention because if ALL functions can suddenly return nil (because a different mod removed the object to which I have a reference to), this will catch many mods by surprise, and is a bug factory. Please add a sentence somewhere that says something like: “Note: All these functions return `nil` and do nothing if the ObjecRef is invalid” or something like this.

There's another question about `remove`: Is it guaranteed that none of the object's callbacks, especially `on_step` is called after `remove`? I remember in Ye Olden Times, this sometimes happened after removing an object, `on_step` was still called a couple of times. But I didn't check lately.

I propose to add an `is_valid` function (maybe in a different PR) that returns `false` for invalid ObjectRefs, because it is useful for more reliable Lua code. This can allow you to skip some processor-heavy functions, or to avoid dozens of `nil` checks for every objectref function.

> Possible bugs where?

Uh, forget it. My question has been answered.

**Comment by Wuzzy2** [comment]
To clarify: I propose to add the “functions return `nil` note in a more prominent place, like at the top of the “ObjectRef” method section, since that's a huge gotcha. Putting this in the `remove` section is correct, but not enough IMO.
Thanks.

**Comment by sfan5** [comment]
> (because a different mod removed the object to which I have a reference to)

But here's the thing. If that happens either you or the other mod has already made the fundamental mistake of storing ObjectRefs.

> Is it guaranteed that none of the object's callbacks, especially on_step is called after remove?

Yes.

> I propose to add an is_valid function (maybe in a different PR) that returns false for invalid ObjectRefs, because it is useful for more reliable Lua code. [..]

I consider such defensive programming unnecessary and adding such a method could in fact cause the opposite. Unless you violate the scoping rules for ObjectRefs, there is absolutely <b>no reason</b> to test if an ObjectRef is valid because you can rely on it being valid.
If you're afraid of other mods messing up, set `deprecated_lua_api_handling = log` and tell the author of that other mod that he sucks.

**Comment by Wuzzy2** [comment]
> But here's the thing. If that happens either you or the other mod has already made the fundamental mistake of storing ObjectRefs.

The documentation doesn't really say that this isn't allowed, right? :P

> the scoping rules for ObjectRefs

Today I heard for the first time that there are scoping rules. :D I always suspected something like that, but I can't remember it being written out explicitly anywhere, and I can't find it with a quick Ctrl-F in `lua_api.txt`.

Lua API docs are suspiciously sparse about what you can (and can't) do with ObjectRefs.

**Comment by sorcerykid** [comment]
I have to agree with Wuzzy. To suggest that someone made the "fundamental mistake of xyz" when xyz isn't officially documented anywhere doesn't seem fair. I remember the first time I learned about this was on a random forum reply from Rubenwardy. People can't reasonably be expected to know every caveat of a complex system, only what they are told :P

**Comment by rubenwardy** [comment]
I was actually wrong, and just repeating what I was told. Storing objectrefs isn't that bad as they are gracefully cleared and won't have segfaults. However, you need to account for invalid objectrefs and they won't become valid again if the object is emerged

**Comment by sfan5** [comment]
> won't have segfaults 

Maybe not inside the ObjectRef code itself, but breaking implicit assumptions could certainly still segfault the engine until now (#9387) or 2016 (b8aed9dfc2ae5e1c877707c856b061ee76b0e1c4) or 2015 (ce8a9ed94b7a3a4cbce04bee27bf8d4167ac6161).

This is also not the case for [anonymous ObjectRefs](https://github.com/minetest/minetest/blob/stable-5/src/script/cpp_api/s_base.cpp#L381), though I couldn't find where these are used or needed.
But if such a ref is kept by Lua beyond the lifetime of the SAO, the engine will crash because it's working on memory that has been freed.

> However, you need to account for invalid objectrefs

There is no API function for this (though several fulfill equivalent purpose) and not having one is preferable IMO.
It is much less error-prone to work with ObjectRefs inside the "intended" scope, you will never need to wonder if the object you are working on still exists.

Unfortunately some mods (such as advtrains) are forced to touch internals like `core.luaentities` to get their functionality working. Maybe the guarantees the engine makes about ObjectRefs could use some revising.

**Comment by Wuzzy2** [comment]
So what *is* a valid scope for ObjectRefs? I'm not sure what you even mean. Do you mean “scope” as in Lua?
What's an anonymous ObjectRef?
And what are the guarantees that the engine makes about ObjectRefs?
Are there other important facts about ObjectRefs, that no modders knows yet?

Please answer the questions in `lua_api.txt`. :-)

> Unfortunately some mods (such as advtrains) are forced to touch internals like core.luaentities to get their functionality working.

But `minetest.luaentities` is officially documented, however, and `core==minetest` AFAIK. So it's not really internal. Documentation also does not warn about it being for “internal use only”. If this table is _meant_ to be internal, documentation needs fixing.

**Comment by sorcerykid** [comment]
> It is much less error-prone to work with ObjectRefs inside the "intended" scope, you will never need to wonder if the object you are working on still exists.

^ This statement right here. Please add that to the documentation. It would be extremely beneficial for modders to know. Full disclosure and all.

**Comment by sfan5** [comment]
> But minetest.luaentities is officially documented [...]. Documentation also does not warn about it being for “internal use only”.

Hmm, this is kind of a problem. `core.luaentities` and `core.object_refs` are there, documented and there are <i>some</i> valid usescases for touching it.

Maybe this stance on ObjectRef reuse is not viable (and not necessary anyway since #9390).

**Comment by sfan5** [comment]
<hr/>

Removed the `ObjectRef` "misuse" tracking and instead added some advice to lua_api.txt

**Comment by sorcerykid** [comment]
Very niiiice verbage! Thanks so much!

**Comment by Wuzzy2** [comment]
> Instead, `ObjectRefs` should be "let go" of as soon as control is returned from
Lua back to the engine.

Just to make sure I understand this right: As long I use an ObjectRef inside the same callback (`minetest.register_on_whatever`), it's safe to use (until it's `remove`d).
As soon I'm outside of this callback, I should no longer trust this ObjectRef.
Is this correct?

**Comment by sfan5** [comment]
Assuming "inside" and "outside" mean the same thing I am thinking of, then yes.

**Comment by Wuzzy2** [comment]
Great!

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -1063,6 +1063,9 @@ int ObjectRef::l_get_luaentity(lua_State *L)
 int ObjectRef::l_is_player_connected(lua_State *L)
 {
 	NO_MAP_LOCK_REQUIRED;
+	// This method was once added for a bugfix, but never documented
+	log_deprecated(L,
+		"is_player_connected is undocumented and may be removed soon");
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: `and will be removed in a future release`

`may` does not motivate enough for actions.

File: src/serverenvironment.h [hunk 2]
```diff
@@ -358,6 +358,7 @@ class ServerEnvironment : public Environment
 
 	RemotePlayer *getPlayer(const session_t peer_id);
 	RemotePlayer *getPlayer(const char* name);
+	const std::vector<RemotePlayer*> getPlayers() const { return m_players; }
```
Comment (SmallJoker) [inline src/serverenvironment.h]: `std::vector<RemotePlayer *>`

**Comment by TheTermos** [comment]
> However, doing this is **NOT** recommended as there is (intentionally) no method to test if a previously acquired `ObjectRef` is still valid.

@sfan5 Why is that? Why intentionally?
The only other way for one object to keep track of another I can think of is to save some data on that object and call get_objects_inside_radius repeatedly, which is as ugly as it gets.

A function to check if an object is still active would be extremely useful, but because it doesn't exist currently my hack is calling get_yaw().
Another useful thing would be get_object_by_id, probably difficult to implement because saos don't do IDs.

But there's nothing, so ObjectRefs will keep being saved, recommended or not.

**Comment by sfan5** [comment]
(This answer focuses on LuaEntities, it should be obvious why there is zero need to store player ObjectRefs.)

> Why is that?

The reason is that objects can unloaded and loaded again at any point, so your ObjectRef will be invalid despite the "same" object still existing. This may or may not be what you want and will come to a surprise on unsuspecting modders who store ObjectRefs.

> Why intentionally?

My opinion is that until the engine provides a way to properly keep track of another object, working around this by storing ObjectRefs should be discouraged.

> The only other way for one object to keep track of another I can think of is [...]

There is [another way](http://git.bananach.space/advtrains.git/tree/advtrains/wagons.lua?id=1bb1d825f46af3562554c12fba35a31b9f7973ff#n980) actually. 

> Another useful thing would be get_object_by_id, probably difficult to implement because saos don't do IDs.

You can index `core.object_refs` by object ID just fine, not sure where you'd get the ID or why you would want to do this though.

**Comment by TheTermos** [comment]
> > Why intentionally?
>
> My opinion is that until the engine provides a way to properly keep track of another object, working around this by storing ObjectRefs should be discouraged.

You wrote that it is intentional that there's no method to test if an ObjectRef is valid.

I don't mind if storing ObjectRefs is discouraged, but taking effort not to provide a method to test the validity is another story.
Such a method is badly needed, considering there's nothing else to get the job done efficiently and the proper way of keeping track of other objects may never happen.

**Comment by sfan5** [comment]
> but taking effort not to provide a method to test the validity is another story.

Providing such a method would encourage storing ObjectRefs and also be a commitment to support that API in the future.
There are 100% reliable workarounds by checking the return value of getters that normally never return nil (`get_yaw`, `get_armor_groups`, `get_velocity`, `get_luaentity`, ...).

**Comment by TheTermos** [comment]
> There are 100% reliable workarounds by checking the return value of getters that normally never return nil (`get_yaw`, `get_armor_groups`, `get_velocity`, `get_luaentity`, ...).

Yes, that what I've been using, but checking object validity isn't their purpose and that behavior can change anytime on other merits.

Anyway I don't see why there seems to be such a problem, checking validity is the first thing one learns when working with stored refs, that's no rocket science.

**Comment by sorcerykid** [comment]
>but checking object validity isn't their purpose

Hmm, I'd argue that the get_luaentity() method would be a perfectly suitable and even coherent way to check whether an object is valid or not. If you can't get the LuaEntity, then the object must not exist.

**Pull request merged**